### DISCUSSION TITLE
Warning: Expected server HTML to contain a matching <div>

### DISCUSSION THREAD

**Discussion by BernardA** [vercel/next.js#15021] [Help]
I have the above show on console and the whole layout goes askew.

It does not seem to happen in all routes, even though the component in question is in the header and that it is shown only when user is logged in. 

I did see some issues in SO ( [like this one](https://stackoverflow.com/questions/58293542/next-js-warning-expected-server-html-to-contain-a-matching-a-in-div-how-to) ), and I understand this is caused by a difference in-between server side and client side rendering but I am not sure how to fix it or even if it isn't a Nextjs issue. 

> react-dom.development.js:88 Warning: Expected server HTML to contain a matching <div> in <div>.
    in div (at statusLoggedMenu.js:81)
    in StatusLoggedMenu (created by WithStyles(StatusLoggedMenu))
    in WithStyles(StatusLoggedMenu) (at status.js:27)
    in div (at status.js:26)
    in div (at status.js:22)
    in Status (created by ConnectFunction)
    in ConnectFunction (created by withCookies(Connect(Status)))
    in withCookies(Connect(Status)) (created by Context.Consumer)
    in withCookies(Connect(Status)) (at header.js:136)
    in div (at header.js:123)
    in div (at header.js:115)
    in header (at header.js:114)
    in Header (at layout.js:15)
    in div (at layout.js:14)
    in div (at layout.js:13)
    in Layout (at _app.js:67)
    in ThemeProvider (at _app.js:65)
    in MyApp (created by withReduxSaga(MyApp))
    in withReduxSaga(MyApp) (created by withRedux(withReduxSaga(MyApp)))
    in Provider (created by withRedux(withReduxSaga(MyApp)))
    in withRedux(withReduxSaga(MyApp))
    in ErrorBoundary (created by ReactDevOverlay)
    in ReactDevOverlay (created by Container)
    in Container (created by AppContainer)
    in AppContainer
    in Root

The full app is available [in this repository](https://github.com/BernardA/quiamo-next). 
The specific component is at `/src/components/statusLoggedMenu

**Comment by thekcsam** [comment]
HI @BernardA, did you manage to solve this?

**Comment by BernardA** [comment]
I noticed a few things: 
- The root cause for this: I had a few cases of client side only output that was not properly wrapped in
 `{typeof window !== 'undefined' ? (`
 conditional. That will also cause this issue. So make sure any output that is not server side is wrapped as above. 
- in development, once you have an error (any unrelated error ) this issue will start to pop up. In that case cleaning the browser cache reset things and this issue goes away. 
- I still do have very rare instances of  this issue that pop up in production. I still haven't found what is causing that.

**Reply by meotimdihia** [reply]
Work for me.

**Reply by meotimdihia** [reply]
The better way to fix this problem: 
https://github.com/vercel/next.js/discussions/17443#discussioncomment-87097

**Comment by meotimdihia** [comment]
This snippet solved the problem and avoid the scroll restoration bug:

```javascript
let mounted = false
export default function Loading() {
  useEffect(() => {
    mounted = true
  }, [])
  
  return (
      {mounted && <div>Run on client only</div>}
  )
 }
 ```

**Reply by filipe-costa** [reply]
I am glad I found this fix. Thanks!

**Reply by meotimdihia** [reply]
Take a look at my newest solution: https://github.com/vercel/next.js/discussions/17443#discussioncomment-739583

**Reply by filipe-costa** [reply]
I already had implemented a similar hook in my codebase, so I just imported it, and it makes total sense it works the way it does.
